pkg/registries/registries.go

```diff
@@ -35,79 +36,182 @@ func ScopeIsNestedInsideScope(subScope, superScope string) bool {
+	for _, idms := range idmsRules {
+		for i := range idms.Spec.ImageDigestMirrors {
+			set := idms.Spec.ImageDigestMirrors[i]
+			digestMirrorSets.addMirrorSets(set.Source, set.MirrorSourcePolicy, set.Mirrors)
+		}
+	}
+	return mergedMirrorsItems(digestMirrorSets)
+}
+
+func mergedICSPMirrorSets(icspRules []*apioperatorsv1alpha1.ImageContentSourcePolicy) ([]mergedMirrorsItem, error) {
+	repositoryMirrorSets := newMirrorSets()
+	for _, icsp := range icspRules {
+		for i := range icsp.Spec.RepositoryDigestMirrors {
+			set := icsp.Spec.RepositoryDigestMirrors[i]
+			imgMirrors := []apicfgv1.ImageMirror{}
 			for _, m := range set.Mirrors {
-				if m == source {
-					sourceInGraph = true
-					break
-				}
+				imgMirrors = append(imgMirrors, apicfgv1.ImageMirror(m))
 			}
-			if !sourceInGraph {
-				// The build of mirrorSets guarantees len(set.Mirrors) > 0.
-				topoGraph.AddEdge(set.Mirrors[len(set.Mirrors)-1], source)
+			// leave MirrorSourcePolicy blank, it will follow the default AllowContactingSource
+			repositoryMirrorSets.addMirrorSets(set.Source, "", imgMirrors)
+		}
+	}
+	return mergedMirrorsItems(repositoryMirrorSets)
+}
+
+func sharedMergedMirrors(source string, ds *[][]string) ([]string, error) {
+	topoGraph := newTopoGraph()
+	for _, mirrors := range *ds {
+		for i := 0; i+1 < len(mirrors); i++ {
+			topoGraph.AddEdge(mirrors[i], mirrors[i+1])
+		}
+		sourceInGraph := false
+		for _, m := range mirrors {
+			if string(m) == source {
+				sourceInGraph = true
+				break
 			}
-			// Every node in topoGraph, including source, is implicitly added by topoGraph.AddEdge (every mirror set contains at least one non-source mirror,
-			// so there are no unconnected nodes that we would have to add separately from the edges).
 		}
-		sortedRepos, err := topoGraph.Sorted()
-		if err != nil {
-			return nil, err
+		if !sourceInGraph {
+			// mirrorSets.addMirrorSets guarantees len(set.Mirrors) > 0.
+			topoGraph.AddEdge(mirrors[len(mirrors)-1], source)
 		}
-		if sortedRepos[len(sortedRepos)-1] == source {
-			// We don't need to explicitly include source in the list, it will be automatically tried last per the semantics of sysregistriesv2. Mirrors.
-			sortedRepos = sortedRepos[:len(sortedRepos)-1]
+		// Every node in topoGraph, including source, is implicitly added by topoGraph.AddEdge (every mirror set contains at least one non-source mirror,
+		// so there are no unconnected nodes that we would have to add separately from the edges).
+	}
+	sortedRepos, err := topoGraph.Sorted()
+	if err != nil {
+		return nil, err
+	}
+	if sortedRepos[len(sortedRepos)-1] == source {
+		// We don't need to explicitly include source in the list, it will be automatically tried last per the semantics of sysregistriesv2. Mirrors.
+		sortedRepos = sortedRepos[:len(sortedRepos)-1]
+	}
+	return sortedRepos, nil
+}
+
+func updateRegistry(mergedMirrorsItems []mergedMirrorsItem, pullFromMirror string, getRegistryEntry func(scope string) *sysregistriesv2.Registry) {
```

Comment:
```suggestion
func {addMirrorsToRegistries,configureMirrors}(mergedMirrorsItems []mergedMirrorsItem, pullFromMirror string, getRegistryEntry func(scope string) *sysregistriesv2.Registry) {
```

`updateRegistry` does not say much about what is being updated.